scripts/pylib/display-twister-harness/camera_shield/uvc_core/camera_controller.py

```diff
@@ -0,0 +1,273 @@
+                if current_score > best_focus_score:
+                    best_focus_score = current_score
+                if i % 20 == 0:  # Periodically adjust focus
+                    self.auto_focus(current_score < 0.5)
+        self.auto_white_balance(True)
+        self.print_settings()
+
+    def get_frame(self):
+        ret, frame = self.cap.read()
+        if ret:
+            frame = cv2.cvtColor(frame, cv2.COLOR_BGR2RGB)
+        return ret, frame
+
+    def show_frame(self, frame):
+        cv2.imshow(f"Camera Preview (Device {self.device_id})", frame)
+        cv2.waitKey(1)
+
+    def auto_focus(self, enable: bool):
+        """Auto focus control
+
+        Args:
+            enable: True to enable auto focus, False to disable
+        """
+        self.cap.set(cv2.CAP_PROP_AUTOFOCUS, 1 if enable else 0)
+
+    def smart_focus(self, frame):
+        """Smart focus algorithm based on image sharpness
+
+        Args:
+            frame: Current video frame
+
+        Returns:
+            Focus score value (0-1)
+        """
+        if frame is None:
+            return 0
+
+        # Convert to grayscale
+        gray = cv2.cvtColor(frame, cv2.COLOR_RGB2GRAY)
+
+        # Calculate Laplacian variance as sharpness metric
+        laplacian = cv2.Laplacian(gray, cv2.CV_64F)
+        focus_score = laplacian.var() / 1000  # Normalized to 0-1 range
+
+        # Adjust focus based on sharpness
+        if focus_score < 0.3:  # Image is blurry
+            self.auto_focus(True)
+        elif focus_score > 0.7:  # Image is sharp
+            self.auto_focus(False)
+
+        return min(max(focus_score, 0), 1)  # Ensure value is within 0-1 range
+
+    def auto_white_balance(self, enable):
+        """Enable/disable auto white balance with algorithm optimization"""
+        if enable:
+            self._original_wb = self.cap.get(cv2.CAP_PROP_WB_TEMPERATURE)
+
+            self.cap.set(cv2.CAP_PROP_AUTO_WB, 1)
+
+            # Sample frames and analyze white balance
+            wb_scores = []
+            for _ in range(10):  # Sample 10 frames for better accuracy
+                ret, frame = self.cap.read()
+                if not ret:
+                    break
+
+                # Convert to LAB color space and analyze A/B channels
+                lab = cv2.cvtColor(frame, cv2.COLOR_BGR2LAB)
+                a_channel, b_channel = lab[:, :, 1], lab[:, :, 2]
+
+                # Calculate white balance score (lower is better)
+                wb_score = np.std(a_channel) + np.std(b_channel)
+                wb_scores.append(wb_score)
+
+                # If score improves significantly, keep current WB
+                if len(wb_scores) > 1 and wb_score < min(wb_scores[:-1]) * 0.9:
+                    break
+        else:
+            if hasattr(self, "_original_wb"):
+                self.cap.set(cv2.CAP_PROP_AUTO_WB, 0)
+                self.cap.set(cv2.CAP_PROP_WB_TEMPERATURE, self._original_wb)
+                # 读取一帧确保设置生效
```

Comment:
I'd love to learn a bit more Chinese, and it's possible to guess what it says with the context, but the guidelines invites documentation to be done in English as much as possible:
https://docs.zephyrproject.org/latest/contribute/contributor_expectations.html#pr-requirements

Whether comments are part of the documentation or not is a blurry limit, I admit. :)